src/rust/engine/src/rule_graph.rs

```diff
@@ -251,115 +322,472 @@ impl<'t> GraphMaker<'t> {
+  /// it exists (it may not, because we're searching for sets of legal parameters in the powerset
+  /// of all used params).
+  ///
+  /// If an ambiguity is detected in rule dependencies (ie, if multiple rules are satisfiable for
+  /// a single dependency key), fail with a Diagnostic.
+  ///
+  fn choose_dependencies<'a>(
+    available_params: &ParamTypes,
+    deps: &'a Vec<(SelectKey, Vec<Entry>)>,
+  ) -> Result<Option<Vec<(&'a SelectKey, &'a Entry)>>, Diagnostic> {
+    let mut combination = Vec::new();
+    for (key, input_entries) in deps {
+      let satisfiable_entries = input_entries
+        .iter()
+        .filter(|input_entry| {
+          Self::used_params(key, input_entry)
+            .iter()
+            .all(|p| available_params.contains(p))
+        })
+        .collect::<Vec<_>>();
+
+      let chosen_entries = Self::choose_dependency(satisfiable_entries);
+      match chosen_entries.len() {
+        0 => {
+          return Ok(None);
+        }
+        1 => {
+          combination.push((key, *chosen_entries.last().unwrap()));
+        }
+        _ => {
+          return Err(Diagnostic::ambiguous(available_params, key, chosen_entries));
+        }
+      }
+    }
+
+    Ok(Some(combination))
+  }
+
+  fn choose_dependency<'a>(satisfiable_entries: Vec<&'a Entry>) -> Vec<&'a Entry> {
+    if satisfiable_entries.is_empty() {
+      // No source of this dependency was satisfiable with these Params.
+      return vec![];
     }
-    result
+
+    // Prefer a Singleton, then a Param, then the non-ambiguous rule with the smallest set of
+    // input Params.
+    // TODO: We should likely prefer Rules to Params.
+    if satisfiable_entries.len() == 1 {
+      satisfiable_entries
+    } else if let Some(singleton) = satisfiable_entries.iter().find(|e| match e {
+      &Entry::Singleton { .. } => true,
+      _ => false,
+    }) {
+      vec![*singleton]
+    } else if let Some(param) = satisfiable_entries.iter().find(|e| match e {
+      &Entry::Param(_) => true,
+      _ => false,
+    }) {
+      vec![*param]
+    } else {
+      // Group by the simplified version of each rule: if exactly one, we're finished. We prefer
+      // the non-ambiguous rule with the smallest set of Params, as that minimizes Node identities
+      // in the graph and biases toward receiving values from dependencies (which do not affect our
+      // identity) rather than dependents.
+      //
+      // TODO: There has to be a more efficient way to group by Entry.
+      let mut rules_by_kind: HashMap<EntryWithDeps, (usize, &Entry)> = HashMap::new();
+      for satisfiable_entry in satisfiable_entries.iter() {
+        match satisfiable_entry {
+          &Entry::WithDeps(ref wd) => {
+            rules_by_kind
+              .entry(wd.simplified(BTreeSet::new()))
+              .and_modify(|e| {
+                // TODO: Param set size isn't sufficient to fully differentiate alternatives: would
```

Comment:
Why do we prefer larger not smaller?

Reply:
This comment is stale: this code is actually prefering the smaller set. As mentioned above, needs some unification.